src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -12,322 +13,87 @@ namespace System
+                // populated data, which is fine since we're splatting the same value for all entries.
+                // There's no need to perform a length check here because we already performed this
+                // check before entering the vectorized code path.
+                Vector<byte>.Zero.StoreUnsafe(ref b, byteLength - (nuint)Vector<byte>.Count);
 
-            if ((((nuint)Unsafe.AsPointer(ref b) - 1) & 4) == 0)
-            {
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i)) = 0;
-                i += 4;
+                // And we're done!
+                return;
             }
 
-            nuint end = byteLength - 16;
-            byteLength -= i; // lower 4 bits of byteLength represent how many bytes are left *after* the unrolled loop
-
-            // We know due to the above switch-case that this loop will always run 1 iteration; max
-            // bytes we clear before checking is 23 (7 to align the pointers, 16 for 1 iteration) so
-            // the switch handles lengths 0-22.
-            Debug.Assert(end >= 7 && i <= end);
+            // If we reached this point, we cannot vectorize this data, or there are too few
+            // elements for us to vectorize. Fall back to an unrolled loop.
+            nuint i = 0;
 
-            // This is separated out into a different variable, so the i + 16 addition can be
-            // performed at the start of the pipeline and the loop condition does not have
-            // a dependency on the writes.
-            nuint counter;
-
-            do
+            // Write 8 elements at a time
+            if (byteLength >= 8)
             {
-                counter = i + 16;
-
-                // This loop looks very costly since there appear to be a bunch of temporary values
-                // being created with the adds, but the jit (for x86 anyways) will convert each of
-                // these to use memory addressing operands.
-
-                // So the only cost is a bit of code size, which is made up for by the fact that
-                // we save on writes to b.
-
-#if TARGET_64BIT
-                Unsafe.As<byte, long>(ref Unsafe.AddByteOffset<byte>(ref b, i)) = 0;
-                Unsafe.As<byte, long>(ref Unsafe.AddByteOffset<byte>(ref b, i + 8)) = 0;
-#else
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i)) = 0;
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i + 4)) = 0;
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i + 8)) = 0;
-                Unsafe.As<byte, int>(ref Unsafe.AddByteOffset(ref b, i + 12)) = 0;
-#endif
-
-                i = counter;
-
-                // See notes above for why this wasn't used instead
-                // i += 16;
+                nuint stopLoopAtOffset = byteLength & ~(nuint)7;
+                do
+                {
+                    Unsafe.WriteUnaligned<ulong>(ref Unsafe.AddByteOffset(ref b, i), 0);
```

Comment:
```suggestion
#if TARGET_64BIT
                     Unsafe.WriteUnaligned<ulong>(ref Unsafe.AddByteOffset(ref b, i), 0);
 #else
                     Unsafe.WriteUnaligned<uint>(ref Unsafe.AddByteOffset(ref b, i), 0);
                     Unsafe.WriteUnaligned<uint>(ref Unsafe.AddByteOffset(ref b, i + 4), 0);
 #endif
```

@EgorBo Actually, your previous revision saved a uop on x86:

https://www.diffchecker.com/nkDFFb34/

But may be better resolved in JIT.

Reply:
Like I said previously - we should not do this sort of unsafe massage for better codegen in general when we own JIT and can fix it there. Anyway, I don't touch it untill I figure out JIT issues I hit in this PR